File: app/src/main/java/cn/com/woong/readhub/ui/widget/newsview/NewsView.java
package cn.com.woong.readhub.ui.widget.newsview;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import com.ajguan.library.EasyRefreshLayout;
import java.util.ArrayList;
import cn.com.woong.readhub.R;
import cn.com.woong.readhub.bean.NewsMo;

/**
 * @author woong
 * Created by wong on 2018/3/12.
 */

public class NewsView extends FrameLayout {
    RecyclerView newsRecyclerView;
    EasyRefreshLayout newsRefreshLayout;

    private NewsAdapter mNewsAdapter;

    public NewsView(@NonNull Context context) {
        this(context, null);
    }

    public NewsView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, -1);
    }

    public NewsView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_news, this);

        newsRecyclerView = findViewById(R.id.news_recycler_view);
        newsRefreshLayout = findViewById(R.id.news_refresh_layout);

        mNewsAdapter = new NewsAdapter(getContext());
        newsRecyclerView.setAdapter(mNewsAdapter);
        newsRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        newsRefreshLayout.addEasyEvent(new EasyRefreshLayout.EasyEvent() {
            @Override
            public void onLoadMore() {
                mOnNewsListener.onNewsLoadMore();
            }

            @Override
            public void onRefreshing() {
                mOnNewsListener.onNewsRefreshing();
            }
        });
    }

    public void updateNews(boolean clear, ArrayList<NewsMo> newsMos) {
        mNewsAdapter.updateNews(clear, newsMos);
        newsRefreshLayout.refreshComplete();
        newsRefreshLayout.loadMoreComplete();
    }

    public void scrollToPosition(int pos) {
        newsRecyclerView.scrollToPosition(pos);
    }

    public void loadMoreComplete() {
        newsRefreshLayout.loadMoreComplete();
    }

    public void refreshComplete() {
        newsRefreshLayout.refreshComplete();
    }

    public interface OnNewsListener {
        void onNewsLoadMore();
        void onNewsRefreshing();
    }

    private OnNewsListener mOnNewsListener;
    public void setOnNewsListener(OnNewsListener listener) {
        mOnNewsListener = listener;
    }
}

File: app/src/main/java/cn/com/woong/readhub/ui/widget/newsview/NewsView.kt
package cn.com.woong.readhub.ui.widget.newsview

import android.content.Context
import android.support.annotation.NonNull
import android.support.annotation.Nullable
import cn.com.woong.readhub.bean.NewsMo
import android.support.v7.widget.LinearLayoutManager
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import cn.com.woong.readhub.R
import com.ajguan.library.EasyRefreshLayout
import kotlinx.android.synthetic.main.view_news.view.*

class NewsView @JvmOverloads constructor(@NonNull context: Context, @Nullable attrs: AttributeSet? = null, defStyleAttr: Int = -1)
    : FrameLayout(context, attrs, defStyleAttr) {

    private var mNewsAdapter: NewsAdapter? = null
    private var mOnNewsListener: OnNewsListener? = null

    init {
        View.inflate(context, R.layout.view_news, this)

        mNewsAdapter = NewsAdapter(context)
        news_recycler_view.adapter = mNewsAdapter
        news_recycler_view.layoutManager = LinearLayoutManager(context)

        news_refresh_layout.addEasyEvent(object : EasyRefreshLayout.EasyEvent {
            override fun onLoadMore() {
                mOnNewsListener!!.onNewsLoadMore()
            }

            override fun onRefreshing() {
                mOnNewsListener!!.onNewsRefreshing()
            }
        })
    }

    fun updateNews(clear: Boolean, newsMos: ArrayList<NewsMo>) {
        mNewsAdapter!!.updateNews(clear, newsMos)
        news_refresh_layout.refreshComplete()
        news_refresh_layout.loadMoreComplete()
    }

    fun scrollToPosition(pos: Int) {
        news_recycler_view.scrollToPosition(pos)
    }

    fun loadMoreComplete() {
        news_refresh_layout.loadMoreComplete()
    }

    fun refreshComplete() {
        news_refresh_layout.refreshComplete()
    }

    interface OnNewsListener {
        fun onNewsLoadMore()
        fun onNewsRefreshing()
    }

    fun setOnNewsListener(listener: OnNewsListener) {
        mOnNewsListener = listener
    }
}


